### ISSUE TITLE
Subset of `metatensor.learn.Dataset` fails at indexing

### ISSUE THREAD

**Issue by frostedoyster** [lab-cosmo/metatensor#521]
When I do
```
for structure, _ in train_dataset:
```
where train_dataset is a `torch.utils.data.Subset` of a `metatensor-learn` `Dataset`, I get
```
ValueError: Index 30 not in dataset
```

**Comment by jwa7** [comment]
@frostedoyster please may you provide an MWE so we can get quicker to the problem? Also is it possible that this was fixed in #500 ?

I've attempted one here, but with no error:

```py
import torch
from metatensor.learn.data import Dataset

dset = Dataset(
    structure=[0, 1, 2, 3, 4, 5],
    y=[0, 1, 2, 3, 4, 5],
)

train_dset, test_dset = torch.utils.data.random_split(dset, [4, 2])

for structure, _ in train_dset:
    print(structure)

for structure, _ in test_dset:
    print(structure)
```
outputs:
```
5
0
2
1
4
3
```

**Comment by jwa7** [comment]
@frostedoyster MWE or shall I close?
